File: 0075-sort-colors/0075-sort-colors.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
//         int n = nums.size();
//         int low = 0, high = n-1, mid = 0;
//         while(mid<=high)
//         {
//             if(nums[mid]==0)
//                 swap(nums[mid++], nums[low++]);
//             else if(nums[mid]==1)
//                 mid++;
//             else
//                 swap(nums[mid],nums[high--]);
//         }
        int lo = 0; 
        int hi = nums.size() - 1; 
        int mid = 0; 

        while (mid <= hi) { 
            switch (nums[mid]) { 

            // If the element is 0 
            case 0: 
                swap(nums[lo++], nums[mid++]); 
                break; 

            // If the element is 1 . 
            case 1: 
                mid++; 
                break; 

            // If the element is 2 
            case 2: 
                swap(nums[mid], nums[hi--]); 
                break; 
            }
        }
        
    }
};

File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        // Initialize three pointers: low and mid at 0, high at the end
        int low = 0, mid = 0, high = nums.length - 1;
        while(mid <= high) {
         // If current element is 0, swap with low and move both low and mid forward
            if(nums[mid] == 0)
            {
                int temp = nums[low];
                nums[low] = nums[mid];
                nums[mid] = temp;
                mid++;
                low++;
            }
            // If current element is 1, just move mid forward
            else if(nums[mid] == 1) {
                mid++;
            }
            // If current element is 2, swap with high and move only high backward

            else {
                int temp = nums[mid];
                nums[mid] = nums[high];
                nums[high] = temp;
                high --;

            }

        }
        
    }
}